cocotb/types/array.py

```diff
@@ -0,0 +1,330 @@
+
+        Returns:
+            A :class:`list` of elements.
+
+        Raises:
+            TypeError: When the type isn't supported.
+            ValueError: When the iterable's values prevent construction into a list of elements.
+        """
+        return list(value)
+
+    @staticmethod
+    def _construct_element(elem: Any) -> Any:
+        """
+        Construct a single element of an array.
+
+        For overriding by subclasses.
+        Used by __setitem__ when setting a single element.
+
+        Args:
+            elem: Any object that the implementation can turn into an array element.
+
+        Returns:
+            An array element.
+
+        Raises:
+            TypeError: When the type isn't supported.
+            ValueError: When the value prevents construction into an element.
+        """
+        return elem
+
+    @staticmethod
+    def _construct_range(rng: Any) -> Range:
+        """
+        Construct the range portion of the array.
+
+        For overriding by subclasses.
+        Used by the constructor when setting the range
+        and by __setitem__ when creating a slice.
+
+        Args:
+            rng: Any object that the implementation can turn into a :class:`Range` object.
+
+        Returns:
+            The :class:`Range` object describing the array's indexing scheme.
+
+        Raises:
+            TypeError: When type isn't supported.
+            ValueError: When the value prevents construction into a range.
+        """
+        if isinstance(rng, Range):
+            return rng
+        raise TypeError("range argument must be of type 'Range'")
+
+    @property
+    def left(self) -> int:
+        """Leftmost index of the array."""
+        return self.range.left
+
+    @property
+    def direction(self) -> str:
+        """``'to'`` if indexes are ascending, ``'downto'`` otherwise."""
+        return self.range.direction
+
+    @property
+    def right(self) -> int:
+        """Rightmost index of the array."""
+        return self.range.right
+
+    @property
+    def length(self):
+        """Length of the array."""
+        return self.range.length
+
+    @property
+    def range(self) -> Range:
+        """:class:`Range` of the indexes of the array."""
+        return self._range
+
+    def __len__(self) -> int:
+        return self.length
+
+    def __iter__(self) -> Iterator[Any]:
+        return iter(self._value)
+
+    def __reversed__(self) -> Iterator[Any]:
+        return reversed(self._value)
+
+    def __contains__(self, item: Any) -> bool:
+        return item in self._value
+
+    def __eq__(self, other: Any) -> bool:
+        sentinel = object()
+        try:
+            it = zip_longest(self, other, fillvalue=sentinel)
+        except TypeError:
+            return NotImplemented
+        for a, b in it:
+            # sentinel check MUST come before element equality check
+            if b == sentinel or a != b:
+                return False
+        return True
```

Comment:
I'm a little confused by this - can you add a docstring describing exactly what you define as equality (and why?).

For instance, is `Array([1, 2, 3]) == (1, 2, 3)` true? How about `Array([1, 2, 3]) == [1, 2, 3]`? Are you sure comparison to arbitrary sequence types is something you want to support given that `list` and `tuple` don't behave that way?

Reply:
Ehhh, I probably shouldn't. It's easy enough to just convert it to an `Array` first.